### DISCUSSION TITLE
new lifecycle (beforeCleanup) proposal

### DISCUSSION THREAD

**Discussion by gabrielmfern** [solidjs/solid#1761] [Ideas]
I have noticed that the onCleanup does not run exactly at the moment that the cleaning up process starts due to children components
needing to be cleaned up before the component itself. This make it kinda hard to work in the situation I am in right now.

I am creating a UI library for Solid and I need to delete the Form's field's value if the field by itself is being cleaned up, but if the field is being cleaned due to the parent being cleaned it should not delete the field's value. 

This is quite simple logic, and to resolve it I propose that we have a new lifecycle that can be used that runs just before the
clean up of the component's children. It can be called `beforeCleanup` or similar. Its usage would be the same as for `onCleanup`.

> This idea is to solve #1757 in a easier way and without using any trickery

**Comment by snnsnn** [comment]
Components do not have lifecycles, these are regular effects, people call them lifecycle for some reason. In overall `befeoreCleanup` serves no purpose, plus Solid is gives you the tool for to address such use cases.

Don't expect out of the box solution for specific problems. There are many ways to address the issue you describe in https://github.com/solidjs/solid/discussions/1757, just read the docs and work out the examples.

**Reply by gabrielmfern** [reply]
Yes, it would serve a purpose, precisely the purpose of solving this problem. We have many out of the box solutions to make people's life easier, why not have another one? Don't really get it.
And we do have things for specific purposes, that are very similar to the one I proposed, such as `createRenderEffect`, why not have a `beforeCleanup` or similar?

Though I disagree, thanks for the terminology explanation and for the comment.

**Reply by gabrielmfern** [reply]
Thinking more about it, the `createRoot` you proposed doesn't really solve it. What other ways can you think of to try solving it? Really don't know, I can only think of adding a `beforeCleanup`.
If `createRoot` solves it, please, show me how because I really need it.

**Reply by snnsnn** [reply]
If you post your code, I will do my best.

**Reply by gabrielmfern** [reply]
Try showing it with the Parent example I posted in #1757

**Reply by snnsnn** [reply]
> Try showing it with the Parent example I posted in #1757

I added an answer there. If I am not mistaken, you are a student. Don't try to be judgemental and don't try to force your logic into a system without having a good understating. Try to be more receptive and change your perspective. There are many design decisions encoded into a system like Solid, small and large, which are products of years of experience and exposure to different thought processes and different needs. When people say no, it is not that they don't want to help but they need to keep the system small and performant. Solid is the most flexible libraries out there which provides you almost all the tools to you need to achieve something. Good luck.

**Reply by gabrielmfern** [reply]
> I added an answer there. If I am not mistaken, you are a student. Don't try to be judgemental and don't try to force your logic into a system without having a good understating. Try to be more receptive and change your perspective. There are many design decisions encoded into a system like Solid, small and large, which are products of years of experience and exposure to different thought processes and different needs. When people say no, it is not that they don't want to help but they need to keep the system small and performant. Solid is the most flexible libraries out there which provides you almost all the tools to you need to achieve something. Good luck.

I am sorry if I seemed judgmental, that was not my intention. Thank you for the answer and your time though I am not a student.

I still do thank that `beforeCleanup` could be useful.

**Reply by snnsnn** [reply]
> I still do thank that `beforeCleanup` could be useful.

There is no event queue, this means there is not an easy way to squeeze it in. You will get a fixed performance hit for something you need rarely because you will be putting additional logic on the critical path.

> I  am not a student.

My apologies.

**Reply by gabrielmfern** [reply]
> There is no event queue, this means there is not an easy way to squeeze it in. You will get a fixed performance hit for something you need rarely because you will be putting additional logic on the critical path.

Actually, looking into the code a bit I think we can add a property `beforeCleanups` in the Owner, and then call these listeners in the cleanNode function just like it is done for `onCleanup`.
